image2tensor/src/image2tensor.rs

```diff
@@ -1,82 +1,97 @@
-pub fn convert_image (
-    path: String,
+pub fn convert_image(
+    path: &String,
     width: u32,
     height: u32,
     precision: TensorType,
-    out_buff: &mut [u8]
-    ) -> i32 {
+    order: ColorOrder,
+) -> Vec<u8> {
     let raw_file = Reader::open(path);
     let raw_file = match raw_file {
         Ok(file) => file,
         Err(e) => {
             println!("Failed to open the file: {:?}", e);
-            return -1;
-        },
+            return vec![];
+        }
     };
 
-    let decoded = raw_file.decode(); //.unwrap();
+    let decoded = raw_file.decode();
     let decoded = match decoded {
         Ok(pixels) => pixels,
         Err(e) => {
             println!("Failed to decode the image: {:?}", e);
-            return -1;
-        },
+            return vec![];
+        }
     };
 
     let dyn_img: DynamicImage = decoded.resize_exact(width, height, image::imageops::Triangle);
-    // Switch from RGB to BGR
-    let bgr_img = dyn_img.to_bgr8();
+    let rgb_bind: ImageBuffer<Rgb<u8>, Vec<u8>>;
+    let bgr_bind: ImageBuffer<Bgr<u8>, Vec<u8>>;
 
     // Get an array of the pixel values
-    let raw_u8_arr: &[u8] = &bgr_img.as_raw()[..];
-    let bytes_per_pixel: usize = get_bytes_per_pixel(precision);
-    let total_bytes: i32 = (raw_u8_arr.len() * bytes_per_pixel) as i32;
-
-    // Create an array to hold the f32 value of those pixels
-    for i in 0..raw_u8_arr.len()  {
-
-        // Read the number as a f32 and break it into u8 bytes
-        let u8_f32: f32 = raw_u8_arr[i] as f32;
-        let u8_bytes = u8_f32.to_ne_bytes();
-
-        for j in 0..bytes_per_pixel {
-            out_buff[(i * bytes_per_pixel) + j] = u8_bytes[j];
+    let raw_u8_arr = match order {
+        ColorOrder::RGB => {
+            rgb_bind = dyn_img.to_rgb8();
+            rgb_bind.as_raw()
         }
+        ColorOrder::BGR => {
+            bgr_bind = dyn_img.to_bgr8();
+            bgr_bind.as_raw()
+        }
+    };
 
-    }
-
-    return total_bytes;
+    return save_bytes(raw_u8_arr, precision);
 }
 
-pub fn calculate_buffer_size (
-    width: u32,
-    height: u32,
-    precision: TensorType) -> usize {
-
+pub fn calculate_buffer_size(width: u32, height: u32, precision: TensorType) -> usize {
     let bytes_per_pixel = get_bytes_per_pixel(precision);
     let pixels: u32 = width * height * 3;
     return pixels as usize * bytes_per_pixel;
 }
 
-fn get_bytes_per_pixel (precision: TensorType) -> usize {
+fn save_bytes(arr: &[u8], tt: TensorType) -> Vec<u8> {
+    let mut out: Vec<u8> = vec![];
+
+    for i in 0..arr.len() {
+        // Split out the bytes based on the TensorType.
+        let bytes: Vec<u8>;
+        bytes = match tt {
+            TensorType::F16 => (arr[i] as f32).to_ne_bytes().to_vec(),
+            TensorType::F32 => (arr[i] as f32).to_ne_bytes().to_vec(),
+            TensorType::U8 => (arr[i] as u8).to_ne_bytes().to_vec(),
+            TensorType::I32 => (arr[i] as i32).to_ne_bytes().to_vec(),
```

Comment:
This seems incorrect: each `match` arm takes a `u8`, pretends it is some other type, and then extracts it as the machine-order bytes of the system. I'm really confused as to why this should be the case and there is no documentation here to explain this. Also, this is the kind of thing that could use some tests to make sure that things work correctly for all the supported precision types.

I would have expected here that this type of conversion would be done by `image` itself. E.g., if we wanted to convert to an 32-bit FP RGB image, then we would use [`to_rgb32f`] and then [`into_vec`]. And so on for the other types. As it is, it seems like we are losing precision (downgrading to an 8-bit value and then casting to a higher-bit value).

[`to_rgb32f`]: https://docs.rs/image/latest/image/enum.DynamicImage.html#method.to_rgb32f
[`into_vec`]: https://docs.rs/image/latest/image/struct.ImageBuffer.html#method.into_vec

For the BGR color order, well, I guess we would have to do some re-arranging of bytes if we move to the latest version of `image`, but I would still expect the `image` crate to help us a bit there.